go/service/usersearch.go

```diff
@@ -188,33 +197,159 @@ func contactSearch(mctx libkb.MetaContext, arg keybase1.UserSearchArg) (res []ke
 	return res, nil
 }
 
+var nonDigits = regexp.MustCompile("[^\\d]")
+
+func imptofuQueryToAssertion(typ keybase1.ImpTofuSearchType, val string) string {
+	switch typ {
+	case keybase1.ImpTofuSearchType_PHONE:
+		return fmt.Sprintf("%s@phone", nonDigits.ReplaceAllString(val, ""))
+	case keybase1.ImpTofuSearchType_EMAIL:
+		return fmt.Sprintf("[%s]@email", strings.ToLower(val))
+	default:
+		panic("invalid keybase1.ImpTofuSearchType enum value")
+	}
+}
+
+func imptofuSearch(mctx libkb.MetaContext, provider contacts.ContactsProvider, imptofuQuery keybase1.ImpTofuQuery) (res *keybase1.APIUserSearchResult, err error) {
+	var emails []keybase1.EmailAddress
+	var phones []keybase1.RawPhoneNumber
+
+	imptofuType, err := imptofuQuery.T()
+	if err != nil {
+		return res, err
+	}
+
+	var queryString string
+
+	switch imptofuType {
+	case keybase1.ImpTofuSearchType_EMAIL:
+		email := imptofuQuery.Email()
+		queryString = string(email)
+		emails = append(emails, email)
+	case keybase1.ImpTofuSearchType_PHONE:
+		phone := keybase1.RawPhoneNumber(imptofuQuery.Phone())
+		queryString = string(phone)
+		phones = append(phones, phone)
+	}
+
+	lookupRes, err := provider.LookupAll(mctx, emails, phones, keybase1.RegionCode(""))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(lookupRes) > 0 {
+		var uids []keybase1.UID
+		for _, v := range lookupRes {
+			if v.Error == "" && !v.UID.IsNil() {
+				uids = append(uids, v.UID)
+			}
+		}
+
+		following, err1 := provider.FindFollowing(mctx, uids)
+		usernames, err2 := provider.FindUsernames(mctx, uids)
+		if err3 := libkb.CombineErrors(err1, err2); err3 != nil {
+			mctx.Warning("Cannot find usernames or tracking info for search results: %s", err3)
+		}
+
+		for _, v := range lookupRes {
+			// Found a resolution
+			if v.Error != "" || v.UID.IsNil() {
+				continue
+			}
+
+			assertionValue := queryString
+			if v.Coerced != "" {
+				assertionValue = v.Coerced
+			}
+			imptofu := &keybase1.ImpTofuSearchResult{
+				CoercedQuery: assertionValue,
+				Resolved:     true,
+				Uid:          v.UID,
+				Assertion:    imptofuQueryToAssertion(imptofuType, assertionValue),
```

Comment:
What is this used for in the resolved case?

Reply:
(ah I see: later you use it for deduplication?)

Reply:
even for resolved case we want the UI to start conversation with assertion, in case the ownership changes between search and createConversation